sklearn/compose/_column_transformer.py

```diff
@@ -1317,12 +1317,31 @@ def get_metadata_routing(self):
 
 
 def _check_X(X):
-    """Use check_array only when necessary, e.g. on lists and other non-array-likes."""
-    if (
-        (hasattr(X, "__array__") and hasattr(X, "shape"))
-        or hasattr(X, "__dataframe__")
-        or sparse.issparse(X)
-    ):
+    """
+    Use `check_array` only when necessary, e.g., on lists and other non-array-likes.
+
+    For sparse matrices, check if they are subscriptable by verifying the presence
+    of the `__getitem__` method. If a sparse matrix is not subscriptable , raise a
+    warning and convert it to the CSR format for compatibility. This ensures
+    efficient and consistent handling of sparse matrices.
+
+    For other input types, such as DataFrame-like objects or array-like structures
+    with `__array__` and `shape` attributes, return them as-is without additional
+    validation to optimize performance.
+
+    Parameters
+    ----------
+    Xs : list of {array-like, sparse matrix, dataframe}
```

Comment:
The input is a single data structure `X`, not a list of such datastructures named `Xs`.